Add `config.storage_client` documentation

georgiy-belyanin opened the pull request:
This patch adds documentation for the `config.storage_client` module.
The module allows one to connect to a remote config.storage cluster and
use it as a key-value storage.

Examples for working both with Tarantool config.storage server API and
Tarantool config.storage client API are provided side by side since they
share quite the same API.

The examples on connecting to a config.storage instance via iproto has
been changed to an example on connecting to multiple config.storage
instances using the config.storage client and interacting with a
config.storage cluster.

georgiy-belyanin committed:
Add `config.storage_client` documentation

This patch adds documentation for the `config.storage_client` module.
The module allows one to connect to a remote config.storage cluster and
use it as a key-value storage.

The examples on connecting to a config.storage instance via iproto has
been changed to an example on connecting to multiple config.storage
instances using the config.storage client and interacting with a
config.storage cluster.

Closes #4734

georgiy-belyanin commented:
Note, that README should also be updated.

https://github.com/tarantool/tarantool/tree/master/doc/examples/config/config-storage

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +360,261 @@ config API
+.. class:: config.storage_client
 
-    Put a value by the specified path.
+The ``config.stoarge_client`` API allows you to connect to a remote replica set :ref:`configured as a Tarantool-based centralized configuration storage <centralized_configuration_set_up_tarantool>` and interact with it as a key-value storage, see Client API examples below.
```
Typo: stoarge -> storage.

georgiy-belyanin replied:
Fixed. Thank you.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +360,261 @@ config API
+.. _config_storage_client_api_reference_connect:
 
-                *   ``revision``: a revision after performing the operation
+.. function:: connect(endpoints[, options])
+
+    Connect to a remote config.storage cluster.
+
+    :param table endpoints: a path to put the value by
+    :param table options: extra options
+
+    :return:    a <config.storage object> object
```
> :param table endpoints: a path to put the value by

Seems incorrect.

> :param table options: extra options

Extra comparing to what?

georgiy-belyanin replied:
Rewritten it a bit. Thank you.
```rst
    :param table endpoints: table of config.storage instance URIs
    :param table options: ``timeout`` - number, connect & call timeout (default: ``nil``, unset)
```

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +360,261 @@ config API
+        :return:    a table containing the following fields:
 
-    Get a value stored by the specified path or prefix.
+                    *   ``recvision``: a revision after performing the operation
```
Typo: recvision -> revision.

georgiy-belyanin replied:
Fixed. Thank you.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +360,261 @@ config API
+            :dedent:
 
-                *   ``revision``: a revision after performing the operation
+        **Client API exapmle:**
```
Typo: exapmle -> example.

georgiy-belyanin replied:
Fixed. Thank you.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -63,19 +63,28 @@ API Reference
+        *   -   :ref:`config.storage_client <config_storage_client_api_reference>`
+            -   Client API for interacting with config.storage clusters
+
+        *   -   :ref:`config.storage_client:connect() <config_storage_client_api_reference_connect>`
+            -   Connect to a remote config.storage cluster
```
Should it be a module function, not an object method?

It seems it also called this way in `myapp_client.lua`.

georgiy-belyanin replied:
Yes. Thank you for noticing. Fixed.

Totktonada commented on doc/reference/reference_lua/config.rst:
It seems the following methods are missed:

* `:close()`
* `:is_closed()`
* `:watch()`
* `:on_disconnect()`
* `:is_connected()`

georgiy-belyanin replied:
Documented them and added `reconnect()` too.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -63,22 +62,52 @@ API Reference
+        *   -   :ref:`config.storage_client.connect() <config_storage_client_api_reference_connect>`
+            -   Connect to a remote config.storage cluster
+
+        *   -   :ref:`<config.storage server>:put() <config_storage_object_api_reference_put>` |br|
```
What is `<config.storage server>`? Why it is called using a colon?

georgiy-belyanin replied:
Fixed. Thank you for noticing.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +380,355 @@ config API
+
+.. class:: config.storage
+
+    If you configure a replicaset as :ref:`Tarantool-based centralized configuration storage <centralized_configuration_storage_tarantool_configure>` by specifying a ``config.storage`` role (see :ref:`configuring roles <configuration_reference_roles>`) the ``config.storage`` object provides ``<config.storage server>`` methods to interact with a key-value storage, see Server API examples below.
```
These are just server API endpoints. We can describe them as an object with methods, but I don't see a reason for that. As for me, it complicates the description.

georgiy-belyanin replied:
I've removed it. Now the documentation on the client and on the server is splitted.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -63,22 +62,52 @@ API Reference
+        *   -   :ref:`config.storage_client.connect() <config_storage_client_api_reference_connect>`
+            -   Connect to a remote config.storage cluster
+
+        *   -   :ref:`<config.storage server>:put() <config_storage_object_api_reference_put>` |br|
+                :ref:`<config.storage client>:put() <config_storage_object_api_reference_put>`
```
I propose to describe the server API and the client API separately. If we mix them in the documentation, they're mixed in the reader's head and it doesn't seem good.

Totktonada replied:
I mean, rearrange the ToC. It is OK to coalesce long descriptions.

georgiy-belyanin replied:
I've rearranged them. Please, take a look when you can.

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +380,355 @@ config API
 
-                *   ``revision``: a revision after performing the operation
+    :param table endpoints: table of config.storage instance URIs
+    :param table options: ``timeout`` - number, connect & call timeout (default: ``nil``, unset)
```
We usually format the options as a list. See `config:get()` description for example.

georgiy-belyanin replied:
Applied.
```diff
+    :param table options: client options to pass. The following options are available:
+
+                           -   ``timeout`` (default: ``nil``, unset) -- connect & call timeout.
```

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +380,355 @@ config API
+.. _config_storage_object_api_reference_put:
 
-.. function:: get(path)
+.. function:: <config.storage server>.put(path, value)
+.. function:: <config.storage client>:put(path, value)
```
How about the `ttl` option? https://github.com/tarantool/enterprise_doc/issues/281

Totktonada commented on doc/reference/reference_lua/config.rst:
How about the `keepalive` method/operation? https://github.com/tarantool/enterprise_doc/issues/282

georgiy-belyanin replied:
I suggest writing about it in a separate PR. It seems it'd cleaner.

Totktonada replied:
It just looks a little extra work and it is unclear, when we return here back next time. Anyway, OK, up to you. Resolving.

Totktonada commented on doc/code_snippets/snippets/centralized_config/instances.enabled/config_storage/myapp.lua:
```diff
@@ -1,6 +1,84 @@
+
+function put_config()
+    local fio = require('fio')
+    local cluster_config_handle = fio.open('../../source.yaml')
```
Nit: It is better to explicitly open the file for reading (`fio.open(<..file..>, {'O_RDONLY'})`). While `O_RDONLY` is actually 0 on Linux, it is not guaranteed by standards AFAIR.

georgiy-belyanin replied:
Thank you for pointing out. Applied and changed the server example too.

```diff
@@ -40,7 +40,7 @@ end
 
 function put_config()
     local fio = require('fio')
-    local cluster_config_handle = fio.open('../../source.yaml')
+    local cluster_config_handle = fio.open('../../source.yaml', {'O_RDONLY'})
     local cluster_config = cluster_config_handle:read()
     local response = storage_client:put('/myapp/config/all', cluster_config)
     cluster_config_handle:close()
diff --git a/doc/code_snippets/snippets/centralized_config/instances.enabled/tarantool_config_storage/myapp.lua b/doc/code_snippets/snippets/centralized_config/instances.enabled/tarantool_config_storage/myapp.lua
index 7df22d09..6d8981a7 100644
--- a/doc/code_snippets/snippets/centralized_config/instances.enabled/tarantool_config_storage/myapp.lua
+++ b/doc/code_snippets/snippets/centralized_config/instances.enabled/tarantool_config_storage/myapp.lua
@@ -1,6 +1,6 @@
 function put_config()
     local fio = require('fio')
-    local cluster_config_handle = fio.open('../../source.yaml')
+    local cluster_config_handle = fio.open('../../source.yaml', {'O_RDONLY'})
     local cluster_config = cluster_config_handle:read()
     local response = config.storage.put('/myapp/config/all', cluster_config)
     cluster_config_handle:close()
```

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +380,355 @@ config API
+
+.. _config_storage_client_api_reference_on_disconnect:
+
+.. function:: <config.storage client>:on_disconnect(func)
+
+Define a trigger for execution when the established connection to a config.storage cluster instance is closed. This also includes the cases when the config.storage client loses connection to a specific instance but automatically re-establishes it to another available one. Execution stops after a config.storage client is explicitly closed, or once the Lua garbage collector removes it.
```
A second call adds another callback (like in net.box) or replaces the existing one? AFAIR, the latter.

georgiy-belyanin replied:
I clarified it

```
Calling this method replaces an old disconnect trigger with a new one if set.
```

Totktonada commented on doc/reference/reference_lua/config.rst:
```diff
@@ -351,178 +380,355 @@ config API
+* ``/path/to/key`` or ``/prefix/to/watch/`` is an event that corresponds to the updates of the specific key or of multiple keys having the specified prefix.
+
+:param string key: a key name of an event to subscribe to
+:param function func:  a callback to invoke when the key value is updated
```
What the callback receives as arguments (in case of `info`, in case of a key, in case of a prefix)?

georgiy-belyanin replied:
Added the information on this 
```rst
:param function func:  a callback to invoke when the key value is updated. The values passed to the callback depend on the passed key and might be one of the following:

                           -   ``event_key``, ``revision`` -- if a path or a prefix is specified.

                           -   ``event_key``, ``info`` -- if the key is ``info``.

```

georgiy-belyanin committed:
Add `config.storage` TTL option

This patch adds a time-to-live (TTL) option to `config.storage` API
methods. This option allows one to put keys that will expire when the
passed time passes.

Closes https://github.com/tarantool/enterprise_doc/issues/281

georgiy-belyanin commented:
@Totktonada  thank you for your detailed review. I am closing the PR since it is made from the fork. Though I've force pushed the changes for you to able to close the comments or to ask for changes.

Suppressed by https://github.com/tarantool/doc/pull/5116.
